source/main/ipc/listenDevices.ts

```diff
@@ -0,0 +1,170 @@
+const getFlatDevices = () => [
+  ...new Set(getDevices().map((d: Device) => flatDevice(d))),
+];
+
+const getDeviceByPaths = (paths) =>
+  listDevices.find((d: Device) => paths.includes(flatDevice(d)));
+
+const lastDevices = new Map();
+
+const addLastDevice = (newDevices: Device[]) =>
+  newDevices.forEach((d) => lastDevices.set(d.path, d));
+addLastDevice(listDevices);
+
+const getPayloadData = (type: 'add' | 'remove', device: Device) => {
+  const descriptor: string = device.path;
+  const deviceModel = identifyUSBProductId(device.productId);
+  return { type, device, deviceModel, descriptor };
+};
+
+// No better way for now. see https://github.com/LedgerHQ/ledgerjs/issues/434
+process.on('exit', () => {
+  if (monitoring) {
+    // redeem the monitoring so the process can be terminated.
+    usbDetect.stopMonitoring();
+  }
+
+  if (timer) {
+    clearInterval(timer);
+  }
+});
+
+let timer;
+
+type Payload = {
+  type: 'add' | 'remove';
+  device: Device;
+  deviceModel: string;
+  descriptor: string;
+};
+
+export const listenDevices = (
+  onAdd: (arg0: Payload) => void,
+  onRemove: (arg0: Payload) => void
+) => {
+  const addEvent = `add:${ledgerUSBVendorId}`;
+  const removeEvent = `remove:${ledgerUSBVendorId}`;
+  let timeout;
+
+  monitor();
+
+  Promise.resolve(getDevices()).then((devices) => {
+    // this needs to run asynchronously so the subscription is defined during this phase
+    for (const device of devices) {
+      onAdd(getPayloadData('add', device));
+    }
+  });
+
+  const poll = () => {
+    log('hid-listen', 'Polling for added or removed devices');
+
+    const currentDevices = getFlatDevices();
+    const newDevices = currentDevices.filter((d) => !lastDevices.has(d));
+
+    if (newDevices.length > 0) {
+      log('hid-listen', 'New device found:', newDevices);
+
+      listDevices = getDevices();
+      onAdd(getPayloadData('add', getDeviceByPaths(newDevices)));
+      addLastDevice(listDevices);
+    } else {
+      log('hid-listen', 'No new device found');
+    }
+
+    const removeDevices = Array.from(lastDevices.keys())
+      .filter((d) => !currentDevices.includes(d))
+      .map((d) => d);
+
+    if (removeDevices.length > 0) {
+      const key = removeDevices[0];
+      const removedDevice = lastDevices.get(key);
+      log('hid-listen', 'Removed device found:', {
+        removeDevices,
+        devices: removedDevice,
+      });
+
+      onRemove(getPayloadData('remove', removedDevice));
+
+      lastDevices.delete(key);
+    } else {
+      log('hid-listen', 'No removed device found');
+    }
+  };
+
+  const debouncedPoll = debounce(poll, usbDebounce);
+
+  const add = (device: usbDetect.Device) => {
+    log('usb-detection', `add: ${deviceToLog(device)}`);
+
+    if (!timeout) {
+      // a time is needed for the device to actually be connectable over HID..
+      // we also take this time to not emit the device yet and potentially cancel it if a remove happens.
+      timeout = setTimeout(() => {
```

Comment:
When you open Cardano App on ledger it will trigger two events in a row: `removed` and `add` 
This timeout is to prevent running the `onAdd` too soon, otherwise you may not be able to retrieve the newly detected device.